stdlib/lists.ncl

```diff
@@ -46,5 +46,33 @@
     any : forall a. (a -> Bool) -> List a -> Bool =
       fun pred l =>
         fold (fun x acc => if pred x then true else acc) l false;
+
+    elem : Dyn -> List -> Bool =
```

Comment:
Why not `forall a. a -> List a -> Bool` ?

Reply:
Good question. This is the type I initially wrote, but actually the equality operator, such as other builtins like `isNum`, doesnn't respect parametricity because it does inspect their arguments. In fact, the implementation clearly depends on the type of each argument. So if you write the proposed polymorphic type, you break the `forall` contract when trying to evaluate the parameters `x` and `elt` in `x == elt`.

For example, in OCaml, the equality does have the type `forall a. a -> a`, but it's a compiler hack, which is regularly called out [[1]](https://blog.janestreet.com/the-perils-of-polymorphic-compare/). On the other hand, in Haskell, the type doesn't lie, since it is constrained by `Eq`, and is not fully polymorphic.

It's even worse currently in Nickel, since such polymorphic values are protected by the contract but don't raise blame when evaluated (they are just special constructs `Wrapped(key, stuff)` preventing further evaluation). Using the proposed polymorphic type, you end up with wrong results without errors, such as :
```
nickel> lists.elem 1 [1,2]
false
```
because everything is wrapped equality doesn't try to inspect wrapped values.

One solution is to cheat and let equality inspect wrapped values, but it feels ad-hoc. Why equality and not `isNum` for example? If every builtin operators can inspect polymorphic values, parametricity doesn't make any sense anymore. I don't see an easy way around it, unless we go through the subtyping route (#244) in which case we could keep the actual types with `Dyn` but without suffering casts in statically typed code, or we can abandon enforcing parametricity in function contracts.

In the end, I think that this is just telling us that the actual type of equality is `Dyn -> Dyn`. It's just typechecked differently (as `_a' -> _a'`) for convenience inside statically typed code, because otherwise you would have to make a type cast to `Dyn` each time you use it in statically typed code, but this is a pragmatic lie.

Reply:
Oh, I see, thanks.

I guess it would theoretically make sense to pass the equality function as a first argument (`elem : forall a. (a -> a -> bool) -> List a -> Bool`), but in that case calling `list.any` directly is as simple

Reply:
> I guess it would theoretically make sense to pass the equality function as a first argument (`elem : forall a. (a -> a -> bool) -> List a -> Bool`), but in that case calling `list.any` directly is as simple

Yes, exactly.